pantsd integration tests fail when run with pantsd enabled

this appears to be a test isolation issue:

```
                     ==================== FAILURES ====================
                      TestPantsDaemonIntegration.test_pantsd_lifecycle_invalidation 
                     
                     self = <pants_test.pantsd.test_pantsd_integration.TestPantsDaemonIntegration testMethod=test_pantsd_lifecycle_invalidation>
                     
                         def test_pantsd_lifecycle_invalidation(self):
                           """Runs pants commands with pantsd enabled, in a loop, alternating between options that
                             should invalidate pantsd and incur a restart and then asserts for pid consistency.
                             """
                     >     with self.pantsd_successful_run_context() as (pantsd_run, checker, workdir):
                     
                     .pants.d/pyprep/sources/a40b9637bd8a73fb48115f1cd8f88d0bf3d4f0bb/pants_test/pantsd/test_pantsd_integration.py:175: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     /opt/ee/python/2.7/lib/python2.7/contextlib.py:17: in __enter__
                         return self.gen.next()
                     .pants.d/pyprep/sources/a40b9637bd8a73fb48115f1cd8f88d0bf3d4f0bb/pants_test/pantsd/test_pantsd_integration.py:89: in pantsd_successful_run_context
                         with self.pantsd_test_context(log_level) as (workdir, pantsd_config, checker):
                     /opt/ee/python/2.7/lib/python2.7/contextlib.py:17: in __enter__
                         return self.gen.next()
                     .pants.d/pyprep/sources/a40b9637bd8a73fb48115f1cd8f88d0bf3d4f0bb/pants_test/pantsd/test_pantsd_integration.py:76: in pantsd_test_context
                         self.assert_success_runner(workdir, pantsd_config, ['kill-pantsd'])
                     .pants.d/pyprep/sources/a40b9637bd8a73fb48115f1cd8f88d0bf3d4f0bb/pants_test/pantsd/test_pantsd_integration.py:103: in assert_success_runner
                         self.run_pants_with_workdir(cmd, workdir, combined_dict(config, extra_config))
                     .pants.d/pyprep/sources/a40b9637bd8a73fb48115f1cd8f88d0bf3d4f0bb/pants_test/pants_run_integration_test.py:318: in assert_success
                         self.assert_result(pants_run, self.PANTS_SUCCESS_CODE, expected=True, msg=msg)
                     .pants.d/pyprep/sources/a40b9637bd8a73fb48115f1cd8f88d0bf3d4f0bb/pants_test/pants_run_integration_test.py:341: in assert_result
                         assertion(value, pants_run.returncode, error_msg)
                     E   AssertionError: /Users/kwilson/dev/pants/pants --no-pantsrc --pants-workdir=/Users/kwilson/dev/pants/.pants.d/tmp/tmplVqKzq.pants.d/.workdir.pants.d --kill-nailguns --print-exception-stacktrace --pants-config-files=/Users/kwilson/dev/pants/.pants.d/tmp/tmplVqKzq.pants.d/.workdir.pants.d/pants.ini kill-pantsd
                     E   returncode: 1
                     E   stderr:
                     E   	/Users/kwilson/dev/pants/src/python/pants/backend/docgen/tasks/confluence_publish.py:21: DeprecationWarning: DEPRECATED: module will be removed in version 1.6.0.dev0.
                     E   	  Use contrib.confluence.tasks.confluence_publish.py instead
                     E   	  deprecated_module('1.6.0.dev0', 'Use contrib.confluence.tasks.confluence_publish.py instead')
                     =============== 5 tests deselected ===============
                     ==== 1 failed, 5 deselected in 11.27 seconds =====
                     
01:00:29 00:17       [coverage]
                     
01:00:30 00:18       [coverage]
                     Name    Stmts   Miss Branch BrPart  Cover
                     -----------------------------------------
                     No data to report.
                     
01:00:30 00:18       [coverage]
                     No data to report.
                     
01:00:31 00:19       [coverage]
                     
                   tests.python.pants_test.pantsd.pantsd_integration                               .....   FAILURE
FAILURE


               Waiting for background workers to finish.
01:00:31 00:19   [complete]
               FAILURE
```

repro:

with pantsd enabled, run:
```
./pants invalidate test tests/python/pants_test/pantsd:pantsd_integration -- -vvs -k test_pantsd_lifecycle_invalidation
```

This was partially fixed by #5262 (ie, most pants integration tests should do the right thing). But this particular test now fails because it still detects the parent process at the end:
```
E           ProcessStillRunning: a pantsd-runner process was detected at PID 23519 (cmdline=['pantsd-runner [./pants test tests/python/pants_test/pantsd: -- -k test_pantsd_lifecycle_invalidation]', '', '', '', '', '', ''])
```

--- a/tests/python/pants_test/pantsd/test_pantsd_integration.py
+++ b/tests/python/pants_test/pantsd/test_pantsd_integration.py
@@ -20,7 +20,7 @@ from pants.pantsd.process_manager import ProcessManager
 from pants.util.collections import combined_dict
 from pants.util.dirutil import touch
 from pants_test.pants_run_integration_test import PantsRunIntegrationTest
-from pants_test.testutils.process_test_util import assert_no_process_exists_by_command
+from pants_test.testutils.process_test_util import no_lingering_process_by_command
 
 
 class PantsDaemonMonitor(ProcessManager):
@@ -85,32 +85,32 @@ def launch_file_toucher(f):
 class TestPantsDaemonIntegration(PantsRunIntegrationTest):
   @contextmanager
   def pantsd_test_context(self, log_level='info'):
-    with self.temporary_workdir() as workdir_base:
-      pid_dir = os.path.join(workdir_base, '.pids')
-      workdir = os.path.join(workdir_base, '.workdir.pants.d')
-      print('\npantsd log is {}/pantsd/pantsd.log'.format(workdir))
-      pantsd_config = {
-        'GLOBAL': {
-          'enable_pantsd': True,
-          # The absolute paths in CI can exceed the UNIX socket path limitation
-          # (>104-108 characters), so we override that here with a shorter path.
-          'watchman_socket_path': '/tmp/watchman.{}.sock'.format(os.getpid()),
-          'level': log_level,
-          'pants_subprocessdir': pid_dir
+    with no_lingering_process_by_command('pantsd-runner'):
+      with self.temporary_workdir() as workdir_base:
+        pid_dir = os.path.join(workdir_base, '.pids')
+        workdir = os.path.join(workdir_base, '.workdir.pants.d')
+        print('\npantsd log is {}/pantsd/pantsd.log'.format(workdir))
+        pantsd_config = {
+          'GLOBAL': {
+            'enable_pantsd': True,
+            # The absolute paths in CI can exceed the UNIX socket path limitation
+            # (>104-108 characters), so we override that here with a shorter path.
+            'watchman_socket_path': '/tmp/watchman.{}.sock'.format(os.getpid()),
+            'level': log_level,
+            'pants_subprocessdir': pid_dir
+          }
         }
-      }
-      checker = PantsDaemonMonitor(pid_dir)
-      self.assert_success_runner(workdir, pantsd_config, ['kill-pantsd'])
-      try:
-        yield workdir, pantsd_config, checker
-      finally:
-        banner('BEGIN pantsd.log')
-        for line in read_pantsd_log(workdir):
-          print(line)
-        banner('END pantsd.log')
+        checker = PantsDaemonMonitor(pid_dir)
         self.assert_success_runner(workdir, pantsd_config, ['kill-pantsd'])
-        checker.assert_stopped()
-        assert_no_process_exists_by_command('pantsd-runner')
+        try:
+          yield workdir, pantsd_config, checker
+        finally:
+          banner('BEGIN pantsd.log')
+          for line in read_pantsd_log(workdir):
+            print(line)
+          banner('END pantsd.log')
+          self.assert_success_runner(workdir, pantsd_config, ['kill-pantsd'])
+          checker.assert_stopped()
 
   @contextmanager
   def pantsd_successful_run_context(self, log_level='info'):
@@ -205,22 +205,21 @@ class TestPantsDaemonIntegration(PantsRunIntegrationTest):
       self.assertIn('Current thread 0x', '\n'.join(read_pantsd_log(workdir)))
 
   def test_pantsd_pantsd_runner_doesnt_die_after_failed_run(self):
-    with self.pantsd_test_context() as (workdir, pantsd_config, checker):
-      # Run target that throws an exception in pants.
-      self.assert_failure(
-        self.run_pants_with_workdir(
-          ['bundle', 'testprojects/src/java/org/pantsbuild/testproject/bundle:missing-files'],
-          workdir,
-          pantsd_config)
-      )
-      checker.await_pantsd()
+    # Check for no stray pantsd-runner prcesses.
+    with no_lingering_process_by_command('pantsd-runner'):
+      with self.pantsd_test_context() as (workdir, pantsd_config, checker):
+        # Run target that throws an exception in pants.
+        self.assert_failure(
+          self.run_pants_with_workdir(
+            ['bundle', 'testprojects/src/java/org/pantsbuild/testproject/bundle:missing-files'],
+            workdir,
+            pantsd_config)
+        )
+        checker.await_pantsd()
 
-      # Check for no stray pantsd-runner prcesses.
-      assert_no_process_exists_by_command('pantsd-runner')
-
-      # Assert pantsd is in a good functional state.
-      self.assert_success(self.run_pants_with_workdir(['help'], workdir, pantsd_config))
-      checker.assert_running()
+        # Assert pantsd is in a good functional state.
+        self.assert_success(self.run_pants_with_workdir(['help'], workdir, pantsd_config))
+        checker.assert_running()
 
   def test_pantsd_lifecycle_invalidation(self):
     """Runs pants commands with pantsd enabled, in a loop, alternating between options that
@@ -264,15 +263,16 @@ class TestPantsDaemonIntegration(PantsRunIntegrationTest):
     # Allow env var overrides for local stress testing.
     attempts = int(os.environ.get('PANTS_TEST_PANTSD_STRESS_ATTEMPTS', 20))
     cmd = os.environ.get('PANTS_TEST_PANTSD_STRESS_CMD', 'help').split()
-    with self.pantsd_successful_run_context('debug') as (pantsd_run, checker, _):
-      pantsd_run(cmd)
-      checker.await_pantsd()
-      for _ in range(attempts):
+
+    with no_lingering_process_by_command('pantsd-runner'):
+      with self.pantsd_successful_run_context('debug') as (pantsd_run, checker, _):
         pantsd_run(cmd)
-        checker.assert_running()
-      # The runner can sometimes exit more slowly than the thin client caller.
-      time.sleep(3)
-      assert_no_process_exists_by_command('pantsd-runner')
+        checker.await_pantsd()
+        for _ in range(attempts):
+          pantsd_run(cmd)
+          checker.assert_running()
+        # The runner can sometimes exit more slowly than the thin client caller.
+        time.sleep(3)
 
   def test_pantsd_aligned_output(self):
     # Set for pytest output display.

--- a/tests/python/pants_test/testutils/process_test_util.py
+++ b/tests/python/pants_test/testutils/process_test_util.py
@@ -23,21 +23,28 @@ def _safe_iter_matching_processes(name):
       pass
 
 
-def _matching_process_set(name):
-  return {p for p in _safe_iter_matching_processes(name)}
+def _make_process_table(processes):
+  line_tmpl = '{0:>7} {1:>7} {2}'
+  proc_tuples = [(p.pid, p.ppid(), ''.join(p.cmdline())) for p in processes]
+  return '\n'.join(
+    [
+      line_tmpl.format('PID', 'PGID', 'CMDLINE')
+    ] + [
+      line_tmpl.format(*t) for t in sorted(proc_tuples)
+    ]
+  )
 
 
 @contextmanager
 def no_lingering_process_by_command(name):
   """Asserts that no process exists for a given command with a helpful error, excluding
   existing processes outside of the scope of the contextmanager."""
-  existing_processes = _matching_process_set(name)
+  before_processes = set(_safe_iter_matching_processes(name))
   yield
-  delta_processes = existing_processes.difference(_matching_process_set(name))
+  after_processes = set(_safe_iter_matching_processes(name))
+  delta_processes = after_processes.difference(before_processes)
   if delta_processes:
-    pids = [p.pid for p in delta_processes]
-    cmdlines = [''.join(p.cmdline()) for p in delta_processes]
     raise ProcessStillRunning(
-      '{} {} processes were detected at PIDS {} (cmdlines={})'
-      .format(len(delta_processes), name, pids, cmdlines)
+      '{} {} processes lingered after tests:\n{}'
+      .format(len(delta_processes), name, _make_process_table(delta_processes))
     )

--- a/tests/python/pants_test/testutils/process_test_util.py
+++ b/tests/python/pants_test/testutils/process_test_util.py
@@ -5,6 +5,8 @@
 from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
                         unicode_literals, with_statement)
 
+from contextlib import contextmanager
+
 import psutil
 
 
@@ -12,14 +14,30 @@ class ProcessStillRunning(AssertionError):
   """Raised when a process shouldn't be running but is."""
 
 
-def assert_no_process_exists_by_command(name):
-  """Asserts that no process exists for a given command with a helpful error."""
+def _safe_iter_matching_processes(name):
   for proc in psutil.process_iter():
     try:
-      cmdline = proc.cmdline()
-      if name in ''.join(cmdline):
-        raise ProcessStillRunning(
-          'a {} process was detected at PID {} (cmdline={})'.format(name, proc.pid, cmdline)
-        )
+      if name in ''.join(proc.cmdline()):
+        yield proc
     except (psutil.NoSuchProcess, psutil.AccessDenied):
       pass
+
+
+def _matching_process_set(name):
+  return {p for p in _safe_iter_matching_processes(name)}
+
+
+@contextmanager
+def no_lingering_process_by_command(name):
+  """Asserts that no process exists for a given command with a helpful error, excluding
+  existing processes outside of the scope of the contextmanager."""
+  existing_processes = _matching_process_set(name)
+  yield
+  delta_processes = existing_processes.difference(_matching_process_set(name))
+  if delta_processes:
+    pids = [p.pid for p in delta_processes]
+    cmdlines = [''.join(p.cmdline()) for p in delta_processes]
+    raise ProcessStillRunning(
+      '{} {} processes were detected at PIDS {} (cmdlines={})'
+      .format(len(delta_processes), name, pids, cmdlines)
+    )

